### PR TITLE
Proposal for effective docker

### PR DESCRIPTION
**Pull request by Thirsty2**
I have created a couple of Dockerfile examples.  The first is Dockerfile.user.  

Currently, it starts FROM symbiflow-ql:1.2.0
It installs make (not absolutely needed, but the version that is already there 
requires conda to be activated, which you may want to do within a Makefile.)
It fixes the permissions on the symbiflow directories via 'chmod 755'
Then it adds a user.
You can run symbiflow tools interactively within this container.

The second, Dockerfile.entrypoint-make is almost identical, but adds
an ENTRYPOINT [ "make" ] 
If you run this container, it runs make on the Makefile in the current directory.
There are come comments/instructions at the bottom of Dockerfile.entrypoint-make
I added a Makefile as an example to show how this works.  I think it is a little simpler
than the Makefile.symbiflow example overall, and it doesn't require any external shell
scripts.  However, I would like it to be simpler still.  Notice that each build target requires activating conda and deactivating conda wrapped around the build command.  I think it may be possible to chain entrypoints such that conda is activated before make runs, which should eliminate this minor complication of the Makefile.   For easy reference, the instructions from the Dockerfile.entrypoint-make are reproduced here:

# Below are some example commands to build and run a docker container
#
# docker build -f Dockerfile.entrypoint-make . -t symbiflow-ql-make 
# 
# In a directory containing a Makefile that runs symbiflow tools:
# docker run -v $(pwd):/home/ic symbiflow-ql-make
#
# The above will execute make within the running container.  Any
# tools called by the makefile will run within the container.
#
# You can run the container with an optional command, which it 
# passes to make.  For instance to run 'make clean' in the container:
#
# # docker run -v $(pwd):/home/ic symbiflow-ql-make clean

Doug

### CODE REVIEW

**Comment by Thirsty2** [comment]
I do have a script that works well as an entrypoint.  It sources the conda.sh, activates conda, then runs make.  It removes the need to do this on every build target line of the makefile, so it simplifies the Makefile considerably.  I am still testing this.  Whem I am finished I can either submit a pull request to this branch, or create a new pull request on the main repo for an additional Makefile.entrypoint-make and sample Makefile.

**Comment by Thirsty2** [comment]
TL;DR The main differences between this PR and the original PR are:
1. Fixed permissions on Symbiflow tools so they can be run by any user.  Did this in parent image (Dockerfile.use-installer).
2. Created Dockerfile.user and Dockerfile.entrypoint-make which wrap symbiflow-ql:latest built from (Dockerfile.use-installer.)
3. Simplified Dockerfile.use-installer-user by wrapping symbiflow-ql:latest as in Dockerfile.user and Dockerfile.entrypoint-make.

Try diffing Dockerfile.user and Dockerfile.entrypoint make.  The differences are very minor.
Try diffing Dockerfile.user and the new Dockerfile.use-installer-user.  The differences are very minor.

Note that the new Dockerfile.use-installer-user doesn't actually use the installer, so is poorly named.  It leverages a container built from the new Dockerfile.use-installer which fixed tool permissions. 

I prefer to have a parent/base image Dockerfile.use-installer that does not add users or entrypoints.  It is easy to add these later as shown in the three example Dockerfile.user, Dockerfile.entrypoint-make, and new Dockerfile.use-installer-user Dockerfiles.  Doing it this way allows users the flexibility to add their own users and entrypoints to match their CI/CD needs or personal preferences.

**Comment by Thirsty2** [comment]
Postnote: I only noticed how similar my conda-make.sh entrypoint script is to the original PR's docker-ql-symbiflow-entrypoint.sh entrypoint script after struggling a while and implementing it, then going back and reviewing the original PR.  I wish I had studied the original more closely, as the solution is simple and would have saved me some time, but I do feel good that we both came up with almost identical entrypoint scripts.  The main difference is that conda-make.sh calls make.

**Comment by Thirsty2** [comment]
I have created a simple sample project that includes a Dockerfile for building a container that runs make.  It contains your fpga_helloworldhw sample, so I hope that's okay.  I modified the Makefile in a similar fashion as described in this thread.  

The project can be found here:
https://github.com/Thirsty2/symbiflow-ql-make

It shows how to build a new container on top of the symbiflow-ql container, adding an entrypoint and a user.  Feel free to use it as you see fit.

**Comment by coolbreeze413** [comment]
@Thirsty2 Thanks for taking the time to simplify the docker approach further, glad to see we weren't that far off in our approaches.
I will try out your changes and we will look to push through a suitable approach that can work well with the other existing projects and the makefile structure of the qorc sdk.
